php/class-customize-concurrency.php

```diff
@@ -14,4 +14,284 @@
+			$is_conflicted = (
+				isset( $saved_setting[ 'timestamp' ], $saved_setting[ 'value' ] )
+				&&
+				$saved_setting['timestamp'] > $timestamps[ $setting_id ]
+				&&
+				$saved_setting['value'] !== $post_values[ $setting_id ]
+			);
+			if ( $is_conflicted ) {
+				$user = get_user_by( 'ID', (int) $saved_setting['author'] );
+				$message = sprintf( 'Value from %s: %s', $user->user_nicename, $saved_setting['value'] );
+				$invalidities[ $setting_id ] = new \WP_Error( 'concurrency_conflict', $message, array( 'their_value' => $saved_setting['value'] ) );
+			}
+		}
+
+		if ( ! empty( $invalidities ) ) {
+			$exported_setting_validities = array_map( array( $wp_customize, 'prepare_setting_validity_for_js' ), $invalidities );
+			$invalid_setting_count = count( $exported_setting_validities );
+			$response = array(
+				'setting_validities' => $exported_setting_validities,
+				'message' => sprintf(
+					_n( 'There is %s conflicting setting.', 'There are %s conflicting settings.', $invalid_setting_count ),
+					number_format_i18n( $invalid_setting_count )
+				),
+			);
+
+			/** This filter is documented in wp-includes/class-wp-customize-manager.php */
+			$response = apply_filters( 'customize_save_response', $response, $this );
+			wp_send_json_error( $response );
+		}
+
+	}
+
+	/**
+	 * Store update history with timestamps.
+	 *
+	 * todo: use snapshots when applicable.
+	 */
+	public function customize_save_after() {
+
+		$post_values = $this->customize_manager->unsanitized_post_values();
+		$setting_ids = array_keys( $post_values );
+		$saved_settings = $this->get_saved_settings();
+
+		$this->suspend_kses();
+		$r = array();
+
+		foreach ( $setting_ids as $setting_id ) {
+			$setting = $this->customize_manager->get_setting( $setting_id );
+			if ( ! $setting ) {
+				continue;
+			}
+
+			$is_update = isset( $saved_settings[ $setting_id ]['post_id'] );
+
+			$post_data = array(
+				'post_type' => self::POST_TYPE,
+				'post_status' => 'publish',
+				'post_title' => $setting_id,
+				'post_name' => $setting_id,
+				'post_content_filtered' => wp_json_encode( $post_values[ $setting_id ] ),
+				'post_date' => current_time( 'mysql', 0 ),
+				'post_date_gmt' => current_time( 'mysql', 1 ),
+				'post_author' => get_current_user_id(),
+			);
+
+			if ( $is_update ) {
+				$post_data['ID'] = $saved_settings[ $setting_id ]['post_id'];
+				$r[] = wp_update_post( wp_slash( $post_data ), true );
+			} else {
+				$r[] = wp_insert_post( wp_slash( $post_data ), true );
+			}
+		}
+
+		$this->restore_kses();
+		// Todo: check $r for errors
+	}
+
+	public function customize_save_response( $response ) {
+		$response['concurrency_session_timestamp'] = strtotime( current_time( 'mysql', true ) );
+		return $response;
+	}
+
+	/**
+	 * Get saved settings from default post storage or from snapshot storage.
+	 *
+	 * Todo: read from snapshots when applicable.
+	 *
+	 * @return array
+	 */
+	function get_saved_settings() {
```

Comment:
@lgedeon let's de-couple this a bit more by letting it take a single argument, `$setting_ids`. Then instead of referencing `$this->customize_manager->unsanitized_post_values()` inside the function, it can then instead refer to `$setting_ids`. And also when calling this function, we can then do:

```php
$post_values = $wp_customize->unsanitized_post_values();
$saved_settings = $this->get_saved_settings( array_keys( $post_values ) );
```

This will make it easier to test and make it more reusable and encapsulated.